src/workerd/api/node/zlib-util.c++

```diff
@@ -878,6 +837,126 @@ void ZlibUtil::CompressionStream<CompressionContext>::FreeForZlib(void* data, vo
 }
+namespace {
+// A RAII wrapper around a compression context class
+template <typename CompressionContext>
+class ContextRAII: public CompressionContext {
+public:
+  using CompressionContext::CompressionContext;
+
+  ~ContextRAII() {
+    static_cast<CompressionContext*>(this)->close();
+  }
+};
+
+template <typename Context>
+static kj::Array<kj::byte> syncProcessBuffer(Context& ctx, GrowableBuffer& result) {
+  do {
+    result.addChunk();
+    ctx.setOutputBuffer(kj::ArrayPtr(result.end(), result.available()));
+
+    ctx.work();
+
+    KJ_IF_SOME(error, ctx.getError()) {
+      JSG_FAIL_REQUIRE(Error, error.message);
+    }
+
+    result.adjustUnused(ctx.getAvailOut());
+  } while (ctx.getAvailOut() == 0);
+
+  return result.releaseAsArray();
+}
+}  // namespace
+
+kj::Array<kj::byte> ZlibUtil::zlibSync(
+    ZlibUtil::InputSource data, ZlibUtil::Options opts, ZlibModeValue mode) {
+  ContextRAII<ZlibContext> ctx;
+
+  auto chunkSize = opts.chunkSize.orDefault(ZLIB_PERFORMANT_CHUNK_SIZE);
+  auto maxOutputLength = opts.maxOutputLength.orDefault(Z_MAX_CHUNK);
+
+  JSG_REQUIRE(Z_MIN_CHUNK <= chunkSize && chunkSize <= Z_MAX_CHUNK, Error, "Invalid chunkSize");
+  JSG_REQUIRE(maxOutputLength <= Z_MAX_CHUNK, Error, "Invalid maxOutputLength");
+  GrowableBuffer result(ZLIB_PERFORMANT_CHUNK_SIZE, maxOutputLength);
+
+  ctx.setMode(static_cast<ZlibMode>(mode));
+  ctx.initialize(opts.level.orDefault(Z_DEFAULT_LEVEL),
+      opts.windowBits.orDefault(Z_DEFAULT_WINDOWBITS), opts.memLevel.orDefault(Z_DEFAULT_MEMLEVEL),
+      opts.strategy.orDefault(Z_DEFAULT_STRATEGY), kj::mv(opts.dictionary));
+
+  auto flush = opts.flush.orDefault(Z_NO_FLUSH);
+  JSG_REQUIRE(Z_NO_FLUSH <= flush && flush <= Z_TREES, RangeError, "invalid flush value");
+
+  auto finishFlush = opts.finishFlush.orDefault(Z_FINISH);
+  JSG_REQUIRE(
+      Z_NO_FLUSH <= finishFlush && finishFlush <= Z_TREES, RangeError, "invalid finishFlush value");
+
+  ctx.setFlush(finishFlush);
+  ctx.setInputBuffer(getInputFromSource(data));
+  return syncProcessBuffer(ctx, result);
+}
+
+void ZlibUtil::zlibWithCallback(
+    jsg::Lock& js, InputSource data, Options options, ZlibModeValue mode, CompressCallback cb) {
+  try {
+    cb(js, kj::none, zlibSync(kj::mv(data), kj::mv(options), mode));
+  } catch (kj::Exception& ex) {
+    auto tunneledError = jsg::tunneledErrorType(ex.getDescription());
+    cb(js, tunneledError.message, kj::none);
+  }
+}
+
+template <typename Context>
+kj::Array<kj::byte> ZlibUtil::brotliSync(InputSource data, BrotliOptions opts) {
+  ContextRAII<Context> ctx(Context::Mode);
+
+  auto chunkSize = opts.chunkSize.orDefault(ZLIB_PERFORMANT_CHUNK_SIZE);
+  auto maxOutputLength = opts.maxOutputLength.orDefault(Z_MAX_CHUNK);
+
+  JSG_REQUIRE(Z_MIN_CHUNK <= chunkSize && chunkSize <= Z_MAX_CHUNK, Error, "Invalid chunkSize");
+  JSG_REQUIRE(maxOutputLength <= Z_MAX_CHUNK, Error, "Invalid maxOutputLength");
```

Comment:
Nit: it would be helpful to expand the error messages here with a bit more detail about why they are invalid. But if we're matching node.js' error messages, then this is fine as is for now.

Reply:
This doesn't match them very well because the NodeJS messages are more structured. I'd like to take a look to see what we can do to improve this later on.

Reply:
Ok, please add a TODO(soon) comment then that will help us keep track and remember